plugin-server/src/main/ingestion-queues/batch-processing/each-batch-ingestion.ts

```diff
@@ -34,9 +34,56 @@ export async function eachBatchIngestion(payload: EachBatchPayload, queue: Inges
         return batches
     }
 
+    await updateLatestEventSentAt(payload, queue)
     await eachBatch(payload, queue, eachMessageIngestion, groupIntoBatchesIngestion, 'ingestion')
 }
 
+async function updateLatestEventSentAt(payload: EachBatchPayload, queue: IngestionConsumer): Promise<void> {
+    // Get the latest sent_at for each token within the batch. We want to
+    // minimise the number of PostgreSQL queries we're adding to the pipeline.
+    const latestSentAtPerToken: Record<number, string> = {}
+    for (const message of payload.batch.messages) {
+        if (message.value) {
+            const event = JSON.parse(message.value.toString())
```

Comment:
So now we'll be json parsing the messages twice? I wonder about the perf impact here vs doing that within EachBatch

Reply:
Yeah not ideal. Perhaps I'll put these values in the message header instead. The reason for not doing this within the inner look was to be able to deduce the latest sent_at within a batch, just to reduce the number of updates we'd do for each team to 1.

Do you think there's an easier way to do this?

Reply:
maybe move it to https://github.com/PostHog/posthog/blob/13bca7138384c9cbaf8975eeafa11b4473d56235/plugin-server/src/main/ingestion-queues/anonymous-event-buffer-consumer.ts#L52 
I suspect we can do https://github.com/PostHog/posthog/blob/13bca7138384c9cbaf8975eeafa11b4473d56235/plugin-server/src/main/ingestion-queues/anonymous-event-buffer-consumer.ts#L93 early there and it also has some error handling, which we don't seem to have here

Reply:
To avoid needing to inspect the message JSON, I have added `team_id`, `api_token` and `sent_at` to the Kafka message headers. Does this work for you?